include/zephyr/spinlock.h

```diff
@@ -29,6 +29,56 @@ extern "C" {
  * @{
  */
 
+/**
+ * @brief Leaves a code block guarded with @ref K_SPINLOCK_GUARDED_SECTION after
+ * releasing the spinlock.
+ *
+ * See @ref K_SPINLOCK_GUARDED_SECTION for details.
+ */
+#define K_SPINLOCK_LEAVE_GUARDED_SECTION continue
+
+/**
+ * @brief Guards a code block with the given spinlock, automatically acquiring
+ * it before executing the code block. The lock will be released either when
+ * reaching the end of the code block or when leaving the block with
+ * K_SPINLOCK_LEAVE_GUARDED_SECTION.
+ *
+ * @details Example usage:
+ *
+ * @code{.c}
+ * K_SPINLOCK_GUARDED_SECTION(&mylock) {
+ *
+ *   ...execute statements with the lock held...
+ *
+ *   if (some_condition) {
+ *     ...release the lock and leave the guarded section prematurely:
+ *     K_SPINLOCK_LEAVE_GUARDED_SECTION;
+ *   }
+ *
+ *   ...execute statements with the lock held...
+ *
+ * }
+ * @endcode
+ *
+ * Be aware that behind the scenes this pattern expands to a for-loop whose body
+ * is executed exactly once:
+ *
+ * @code{.c}
+ * for (k_spinlock_key_t key = k_spin_lock(&mylock); ...; k_spin_unlock(&mylock, key)) {
+ *     ...
+ * }
+ * @endcode
+ *
+ * @note The code block must execute to its end or be left by calling
+ * K_SPINLOCK_LEAVE_GUARDED_SECTION. Otherwise, e.g. if exiting the block with
+ * a break or return statement, the spinlock will not be released on exit.
```

Comment:
Might want to add `goto` to the list, as "linux goto style" error handling is fairly common in Zephyr and is incompatible with this trick too.